bridges/snowbridge/primitives/router/src/inbound/v2.rs

```diff
@@ -0,0 +1,257 @@
+		let mut message_xcm: Xcm<()> = Xcm::new();
+		if message.xcm.len() > 0 {
+			// Decode xcm
+			let versioned_xcm = VersionedXcm::<()>::decode_with_depth_limit(
+				MAX_XCM_DECODE_DEPTH,
+				&mut message.xcm.as_ref(),
+			)
+			.map_err(|_| ConvertMessageError::InvalidVersionedXCM)?;
+			message_xcm = versioned_xcm.try_into().map_err(|_| ConvertMessageError::InvalidXCM)?;
+		}
+
+		log::debug!(target: LOG_TARGET,"xcm decoded as {:?}", message_xcm);
+
+		let network = EthereumNetwork::get();
+
+		// use weth as asset
+		let fee_asset = Location::new(
+			2,
+			[
+				GlobalConsensus(EthereumNetwork::get()),
+				AccountKey20 { network: None, key: WethAddress::get().into() },
+			],
+		);
+		let fee: XcmAsset = (fee_asset.clone(), message.fee).into();
+		let mut instructions = vec![
+			DescendOrigin(PalletInstance(InboundQueuePalletInstance::get()).into()),
+			UniversalOrigin(GlobalConsensus(network)),
+			ReserveAssetDeposited(fee.clone().into()),
+			PayFees { asset: fee },
+		];
+
+		for asset in &message.assets {
+			match asset {
+				Asset::NativeTokenERC20 { token_id, value } => {
+					let token_location: Location = Location::new(
+						2,
+						[
+							GlobalConsensus(EthereumNetwork::get()),
+							AccountKey20 { network: None, key: (*token_id).into() },
+						],
+					);
+					instructions.push(ReserveAssetDeposited((token_location, *value).into()));
+				},
+				Asset::ForeignTokenERC20 { token_id, value } => {
+					let asset_id = ConvertAssetId::convert(&token_id)
+						.ok_or(ConvertMessageError::InvalidAsset)?;
+					instructions.push(WithdrawAsset((asset_id, *value).into()));
+				},
+			}
+		}
+
+		if let Some(claimer) = message.claimer {
+			let claimer = Junction::decode(&mut claimer.as_ref())
+				.map_err(|_| ConvertMessageError::InvalidClaimer)?;
+			let claimer_location: Location = Location::new(0, [claimer.into()]);
+			instructions.push(SetAssetClaimer { location: claimer_location });
+		}
+
+		let mut origin_location = Location::new(2, [GlobalConsensus(network)]);
+		if message.origin == GatewayProxyAddress::get() {
+			// If the message origin is the gateway proxy contract, set the origin to
+			// Ethereum, for consistency with v1.
+			instructions.push(AliasOrigin(origin_location));
+		} else {
+			// Set the alias origin to the original sender on Ethereum. Important to be before the
+			// arbitrary XCM that is appended to the message on the next line.
+			origin_location
+				.push_interior(AccountKey20 { key: message.origin.into(), network: None })
+				.map_err(|_| ConvertMessageError::InvalidOrigin)?;
+			instructions.push(AliasOrigin(origin_location.into()));
+		}
+
+		// Add the XCM sent in the message to the end of the xcm instruction
+		instructions.extend(message_xcm.0);
+
+		let appendix = vec![
+			RefundSurplus,
+			// Refund excess fees to the relayer
+			DepositAsset {
+				assets: Wild(AllOf { id: AssetId(fee_asset.into()), fun: WildFungible }),
+				beneficiary: origin_account_location,
```

Comment:
We should maybe refund to the claimer if its present, and if its not than refund to the relayer. The relayer will get its reward separately.

Reply:
https://github.com/paritytech/polkadot-sdk/pull/6697/commits/85d355e1d8d5d2e4d0f8d4c5a2d580c92611593a